### DISCUSSION TITLE
Composition patterns and practices

### DISCUSSION THREAD

**Discussion by negz** [crossplane/crossplane#3122] [Knowledge base]
Are you using XRs and Compositions in production? Or do you have some patterns and practices you'd like to share? We'd love to hear how you're using it - how would you recommend a new Crossplane user approach building their own library of Compositions?

**Comment by ytsarev** [comment]
Example from AWS friends https://github.com/aws-samples/crossplane-aws-blueprints

**Comment by ksawerykarwacki** [comment]
I am building whole infrastructure/application provisioning using Crossplane (with assist of flux to sync definitions/compositions to management cluster). After three months of work with Crossplane I find it as a brilliant tool but with lots and lots of shortcomings which require workarounds and pretty much wastes teams time and makes compositions messy so be aware that it will be difficult especially if you have more complexed scenarios involving multiple providers. But let's walk through list of my "lessons learned" and few tips (the text below may look like a rant, but it is not. I just love Crossplane and want it to become even better :)):

1. Providers are not created equaly. We are using aws, helm and kubernetes and api is at best inconsistent. You want to patch from an existing secret to your composition? That is great, you can patch any kubernetes object or helm release but cannot do the same with aws objects. Workaround? Use Kubernetes provider pointing to your secret with managementPolicy set to Observe (which by the way is only supported by this specific provider) and patch from the manifest stored in status of that object. After that spend 20 minutes on PR review explaining to your team why this is not janky at all :)

2. Patches can solve most of your issues but very often it requires deep understanding what should be patched and how to force order of operations using patching policy set to Optional/Required. You also have to store a lot of trash information on your composition as there is no other way to get information between resources. Finally the only way to use conditions is to use transforms/map without mapping all possible source values (it will work only on mapped values and break silently on the others). Again, spend 20 minutes explaining that this is not a janky hack :). Furthermore scenarios when you have to merge values from multiple MRs to one field (e.g. config map with json containing s3 bucket connection [account, url, region, secret key]) will get pretty complicated and when it finally works you will probable would like to never look at this part of yaml ever again.

3. If patching is not enough you probably need naming convention especially when you have to pass information between separate compositions (e.g. you provision infrastructure and app instance separately). As mentioned above, some providers support mapping from existing resource, some do not. https://github.com/upbound/platform-ref-aws presents nice way how this can be achieved using simple naming convention (look at the usage of NetworkRef and ClusterRef).

4. There is a lot of repetition and there is no way around it so don't be afraid of duplicated manifests. As there is no way to have conditions or iterate over anything you have to manually support all possible scenarios which results in multiple compositions for each definition. (e.g. EKS cluster definition where sometimes you need one node group and sometimes three, even though you can define array of values/objects on definition, composition cannot render multiple NodeGroup MRs based on the list (unless you use kubernetes/helm provider as a middleman but this will just make things worse than just writing two compositions and duplicate all the common patches).

5. Crossplane behavior is not well documented and you will often think that something stopped working but it is just part of how things work in this land (e.g. problems discussed at https://github.com/crossplane/crossplane/discussions/3116, again multiple providers used together so not just simple scenario that you will find in the examples)

6.  You have to figure out how to handle multiple versions of your compositions. Crossplane doesn't have full CRD versioning supported. You have to use something like variable substitution from Flux or use some string replacement during deploy to provide both dev and prod versions or just use separate clusters if you can afford it. There is no documentation how to introduce changes to your compositions, how providers will handle existing XRs and claims when you change the definition. So there is a lot of things where you have to figure out stuff or pray that someone on slack had the same issue. 

Most of the janky workarounds described above should be solved when we will finally get https://github.com/crossplane/crossplane/issues/2524 and https://github.com/crossplane/crossplane/issues/2099. But until then if you want to use Crossplane buckle up as it will be a bumpy ride :)

Tl;dr: If your solution does not work you probably need more patches or duplicated compositions for each of your scenarios ;)
